Shape and dtype check for FrozenModel

pyu10055 opened the pull request:
Added shape and dtype check for input tensors.

fixed https://github.com/tensorflow/tfjs/issues/370

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/tensorflow/tfjs-converter/150)
<!-- Reviewable:end -->

bileschi commented on src/executor/graph_executor.ts:
```diff
@@ -221,17 +224,37 @@ export class GraphExecutor {
+      if (node.params['dtype'] && node.params['dtype'].value) {
+        util.assert(
+            input.dtype === node.params['dtype'].value as string,
+            `The dtypes of the ${node.name}(${input.dtype}) and` +
```
This error message doesn't quite make sense gramatically:

perhaps:

The dtype of the input is `float32` but `int32` was expected'

pyu10055 commented:
Review status: 0 of 1 LGTMs obtained

---

*[src/executor/graph_executor.ts, line 241 at r1](https://reviewable.io/reviews/tensorflow/tfjs-converter/150#-LEWVKYTupil88h9bAC7-r1-241:-LEjzYsE3qxk4oXVKA-U:b-s9k0rj) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/46d1327bc989b50c666c76dc54582d397294c7ba/src/executor/graph_executor.ts#L241)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

This error message doesn't quite make sense gramatically:

perhaps:

The dtype of the input is `float32` but `int32` was expected'
</blockquote></details>

thanks for the suggestion. updated.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-converter/150)*
<!-- Sent from Reviewable.io -->

dsmilkov commented:
Thank you for adding these! Couple comments

---

Reviewed 1 of 4 files at r1, 2 of 2 files at r2.
Review status: 0 of 1 LGTMs obtained

---

*[src/executor/graph_executor.ts, line 236 at r2](https://reviewable.io/reviews/tensorflow/tfjs-converter/150#-LEtiMLK-JiYIYL4a8pH:-LEtiMLK-JiYIYL4a8pI:b-xhuucl) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/05aece76a67e5f1d369c9271e6b34662b22788b6/src/executor/graph_executor.ts#L236)):*
> ```XML
>                 (dim, index) => shape[index] === -1 || shape[index] === dim);
>         util.assert(
>             match, `${node.name} shape(${input.shape}) must be ${shape}`);
> ```

How about:
The shape of dict['${node.name}'] provided in model.execute(dict) must be [${shape}], but was [${input.shape}]

This will produce an example message:
`The shape of dict['images'] provided in model.execute(dict) must be [1, 32, 32, 3] but was [32, 32, 3].`

Do the same for the dtype below.

---

*[src/executor/graph_executor_test.ts, line 113 at r2](https://reviewable.io/reviews/tensorflow/tfjs-converter/150#-LEthjg-8zx0JQteRoKF:-LEthjg-8zx0JQteRoKG:b-y7qu4z) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/05aece76a67e5f1d369c9271e6b34662b22788b6/src/executor/graph_executor_test.ts#L113)):*
> ```XML
>       });
> 
>       it('should throw exception if inputs dtype do not match graph', () => {
> ```

add unit tests that are calling model.predict() (in addition to model.execute()).

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-converter/150#-:-LEtjL1M8D_9T_sSo6nw:b-fm5ia)*
<!-- Sent from Reviewable.io -->

pyu10055 commented:
thanks for the review, updated accordingly.

---

Review status: 0 of 1 LGTMs obtained

---

*[src/executor/graph_executor.ts, line 241 at r1](https://reviewable.io/reviews/tensorflow/tfjs-converter/150#-LEWVKYTupil88h9bAC7-r1-241:-LEuGoY65P2VAZMKKv3d:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/46d1327bc989b50c666c76dc54582d397294c7ba/src/executor/graph_executor.ts#L241)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

thanks for the suggestion. updated.
</blockquote></details>

Done.

---

*[src/executor/graph_executor.ts, line 236 at r2](https://reviewable.io/reviews/tensorflow/tfjs-converter/150#-LEtiMLK-JiYIYL4a8pH:-LEuGifzCu4BZr9u7S3x:bxrkjs) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/05aece76a67e5f1d369c9271e6b34662b22788b6/src/executor/graph_executor.ts#L236)):*
> The shape of dict['${node.name}'] provided in model.execute(dict) must be [${shape}], but was [${input.shape}]
done

---

*[src/executor/graph_executor_test.ts, line 113 at r2](https://reviewable.io/reviews/tensorflow/tfjs-converter/150#-LEthjg-8zx0JQteRoKF:-LEuI7sn4KaLbly2xhVW:b3l8l3a) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/05aece76a67e5f1d369c9271e6b34662b22788b6/src/executor/graph_executor_test.ts#L113)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

add unit tests that are calling model.predict() (in addition to model.execute()).
</blockquote></details>

the graph executor does not have predict method, the predict method is on the FrozenModel. Unless you think the test needs to be added there.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-converter/150#-:-LEuJNsU-fYrnuwwlSMd:b-d4e2rc)*
<!-- Sent from Reviewable.io -->

dsmilkov commented:
Review status: 0 of 1 LGTMs obtained

---

*[src/executor/graph_executor_test.ts, line 113 at r2](https://reviewable.io/reviews/tensorflow/tfjs-converter/150#-LEthjg-8zx0JQteRoKF:-LEuLRFoC6qy-yf3jk9x:bl0uc37) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/05aece76a67e5f1d369c9271e6b34662b22788b6/src/executor/graph_executor_test.ts#L113)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

the graph executor does not have predict method, the predict method is on the FrozenModel. Unless you think the test needs to be added there.
</blockquote></details>

Since this change has effect on `frozenmodel.predict()` add unit tests there too (new ones, keep these here). these will test public api which is a good thing.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-converter/150)*
<!-- Sent from Reviewable.io -->

pyu10055 commented:
Review status: 0 of 1 LGTMs obtained

---

*[src/executor/graph_executor_test.ts, line 113 at r2](https://reviewable.io/reviews/tensorflow/tfjs-converter/150#-LEthjg-8zx0JQteRoKF:-LEzETQBDfm-vtBsb3bx:b-xql2uh) ([raw file](https://github.com/tensorflow/tfjs-converter/blob/05aece76a67e5f1d369c9271e6b34662b22788b6/src/executor/graph_executor_test.ts#L113)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

Since this change has effect on `frozenmodel.predict()` add unit tests there too (new ones, keep these here). these will test public api which is a good thing.
</blockquote></details>

test added.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-converter/150)*
<!-- Sent from Reviewable.io -->

dsmilkov commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> Nice! Thank you!

---

Reviewed 1 of 4 files at r1, 2 of 2 files at r3, 1 of 1 files at r4.
Review status: :shipit: complete! 1 of 1 LGTMs obtained

---



*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-converter/150#-:-LEzPNqF7W-858Hnk9Q8:b-kcyse1)*
<!-- Sent from Reviewable.io -->
